### ISSUE TITLE
More Cool Stuff with Embedded App Installs

### ISSUE THREAD

**Issue by resistorsoftware** [Shopify/shopify_app#1133]
shopify_app 16.1.0

TLDR: When the App is approved in the admin, and redirects back to the App, App Bridge is not picking this up and ensuring we're all good in the iframe. Why? What causes that? 

Starting with the login screen: https://myapp.test/login
Provide the name of the Shopify store. Get redirected to store install modal. Approve App
Redirect to the App outside the iframe, so App is not embedded. 
Turned JWT ON in the config of the App

So what is the culprit here? 

The console of the App displaying myapp.test shows rendering of the code

    Started GET "/?shop=apptest.myshopify.com" for 127.0.0.1 at 2021-01-11 12:15:45 -0500
    Processing by HomeController#index as HTML
    Parameters: {"shop"=>"apptest.myshopify.com"}
    Shop Load (0.2ms)  SELECT "shops".* FROM "shops" WHERE "shops"."shopify_domain" = $1 LIMIT $2  [["shopify_domain",   "apptest.myshopify.com"], ["LIMIT", 1]]
    Rendering home/index.html.erb within layouts/embedded_app
    Rendered home/index.html.erb within layouts/embedded_app (Duration: 0.0ms | Allocations: 4)
    Rendered layouts/_flash_messages.html.erb (Duration: 0.1ms | Allocations: 84)
    Completed 200 OK in 5ms (Views: 3.3ms | ActiveRecord: 0.2ms | Allocations: 3828)`

So it looks like the App is getting the shop parameter it needs, and rendering out the proper stuff, which ultimately is going to be a React App using Polaris, featuring App Bridge. 

Except App Bridge does not render properly, so this App just renders by itself in the browser, outside the Shopify store, which is not what I expected. So how to debug why App Bridge is bombing? 

The console shows only one failure, and that is a postMessage due to domain screwups. 

   Failed to execute ‘postMessage’ on ‘DOMWindow’: The target origin provided (‘https://apptest.myshopify.com’) does not match the recipient window’s origin (‘https://myapp.test’).

So the browser is sitting there showing off a Polaris App, outside the Shop at:

    https://myapp.test/?shop=apptest.myshopify.com

So what is the cause of this? 

If I login to the shopify store, and choose the App link, it is installed perfect, and everything works perfect, including the JWT tokens and all the code. So to me, this is a problem stemming somehow from the redirect from Shopify, back to the App and getting it to embed in the iframe properly. That is where I am puzzled.

**Comment by resistorsoftware** [comment]
I used some code, which worked to mitigate this, but it seems wrong to have to do this so manually. In the layout I just redirect. 

    <script>
      window.SHOPIFY_API_KEY = "<%= ShopifyApp.configuration.api_key %>";
      window.SHOP_ORIGIN = "<%= @shop_origin %>";
      if (window.parent == window) {
        document.querySelector('#main').innerHTML = '<p>Redirecting to Shopify Admin...</p>';
        window.location = 'https://' + SHOP_ORIGIN + '/admin/apps/' + SHOPIFY_API_KEY;
      }
    </script>

**Comment by paulomarg** [comment]
Hi @resistorsoftware, thank you for the detailed report and workaround. App Bridge should automatically redirect the app to Admin after OAuth completes, but it currently only does that when the environment is set to production.

Could you please retest this with production settings so we can see if the issue still reproduces? If it is still there we will look into fixing it.

**Comment by resistorsoftware** [comment]
How does App Bridge know what production versus development is? What flag is it checking?

**Comment by paulomarg** [comment]
I believe that App Bridge will check whether the `NODE_ENV` variable (as in `process.env.NODE_ENV`) is **not** `development`. It will not force the redirect if the value is `development`.

**Comment by resistorsoftware** [comment]
I am running the App on Rails. So I control RAILS_ENV. Are you saying I should fake out the App Bridge by make a NODE_ENV and fooling it that way? Or is there some way to set this NODE_ENV properly? I was not under the impression NODE_ENV was something I had to deal with in the Shopify App gem.

**Comment by paulomarg** [comment]
You're right, you shouldn't have to manage that flag from the Rails side of things, except when building front-end assets.

In order to help us investigate this further: have you made any changes to the default settings that might affect this, like not using webpacker?

**Comment by resistorsoftware** [comment]
I use Webpacker. It is a standard Rails 6 Shopify App, with Polaris installed, App Bridge, and it uses Webpacker. I just never deal with the concept of Node or its settings. I use Yarn to add/update packages, and that is about it. So how this App Bridge is failing due to the environment variable is not something I have seen before. Usually my Apps just install and work even in localhost scenarios. This JWT configuration TRUE is the first time it has not worked.
